optaplanner-examples/src/main/java/org/optaplanner/examples/coachshuttlegathering/optional/score/CoachShuttleGatheringConstraintProvider.java

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright 2020 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.optaplanner.examples.coachshuttlegathering.optional.score;
+
+import org.optaplanner.core.api.score.buildin.hardsoftlong.HardSoftLongScore;
+import org.optaplanner.core.api.score.stream.Constraint;
+import org.optaplanner.core.api.score.stream.ConstraintFactory;
+import org.optaplanner.core.api.score.stream.ConstraintProvider;
+import org.optaplanner.examples.coachshuttlegathering.domain.Bus;
+import org.optaplanner.examples.coachshuttlegathering.domain.BusStop;
+import org.optaplanner.examples.coachshuttlegathering.domain.Coach;
+import org.optaplanner.examples.coachshuttlegathering.domain.Shuttle;
+import org.optaplanner.examples.coachshuttlegathering.domain.StopOrHub;
+
+import static org.optaplanner.core.api.score.stream.ConstraintCollectors.count;
+import static org.optaplanner.core.api.score.stream.ConstraintCollectors.sum;
+import static org.optaplanner.core.api.score.stream.Joiners.equal;
+
+public class CoachShuttleGatheringConstraintProvider implements ConstraintProvider {
+
+    @Override
+    public Constraint[] defineConstraints(ConstraintFactory constraintFactory) {
+        return new Constraint[] {
+                coachStopLimit(constraintFactory),
+                shuttleCapacity(constraintFactory),
+//                coachCapacity(constraintFactory),
+                transportTime(constraintFactory),
+                shuttleDestinationIsCoachOrHub(constraintFactory),
+                shuttleSetupCost(constraintFactory),
+                distanceFromPrevious(constraintFactory),
+                distanceBusStopToBusDestination(constraintFactory),
+                distanceCoachDirectlyToDestination(constraintFactory)
+        };
+    }
+
+    // ############################################################################
+    // Hard constraints
+    // ############################################################################
+
+    private Constraint coachStopLimit(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(BusStop.class)
+                .filter(busStop -> busStop.getBus() instanceof Coach)
+                .groupBy(busStop -> (Coach) busStop.getBus(), count())
+                .filter((coach, stopCount) -> stopCount > coach.getStopLimit())
+                .penalizeLong("coachStopLimit", HardSoftLongScore.ofUninitialized(0, 1000000, 0),
```

Comment:
Why not `HardSoftLongScore.ofHard(1000000)`?
Similarly in the other constraints.